flink-yarn-tests/src/test/java/org/apache/flink/yarn/YarnTestBase.java

```diff
@@ -212,30 +212,45 @@ public void checkClusterEmpty() {
 	 * Sleep a bit between the tests (we are re-using the YARN cluster for the tests).
 	 */
 	@After
-	public void sleep() throws IOException, YarnException {
-		Deadline deadline = Deadline.now().plus(Duration.ofSeconds(10));
+	public void shutdownYarnClient() {
+		yarnClient.stop();
+	}
 
-		boolean isAnyJobRunning = yarnClient.getApplications().stream()
-			.anyMatch(YarnTestBase::isApplicationRunning);
+	protected void runTest(RunnableWithException test) {
+		Throwable testFailure = null;
+		try {
+			test.run();
+		} catch (Throwable t) {
+			testFailure = t;
```

Comment:
No need to catch `t` if the application shut down is done in a `finally` block.

Reply:
Since the finally block can throw exceptions, without catching the exception here you may end up swallowing the test exception entirely, which would be worse than the current state.

Reply:
No, this would not happen. If the finally block throws an exception it would either throw this exception if the try block has not thrown an exception or add the finally exception as a suppressed exception to the try exception.

Reply:
are you saying that, given the piece of code below, exception "2" would be automatically suppressed?
```
try {
	throw new Exception("1");
} finally {
	throw new Exception("2");
}
```
According to this [SO post](https://stackoverflow.com/questions/19583374/are-exceptions-thrown-in-a-finally-block-added-to-the-list-of-suppressed-excepti) this only works when using try-with-resources.
Thus in order to suppress an exception thrown in the `finally` part we'd still have to catch the test failure exception beforehand.
Which we could of course, but we'd end up with:
```
Throwable testFailure = null;
try {
	test.run();
} catch (Throwable t) {
	testFailure = t;
} finally {
	try {
		Deadline deadline = Deadline.now().plus(Duration.ofSeconds(10));

		boolean isAnyJobRunning = yarnClient.getApplications().stream()
			.anyMatch(YarnTestBase::isApplicationRunning);

		while (deadline.hasTimeLeft() && isAnyJobRunning) {
			try {
				Thread.sleep(500);
			} catch (InterruptedException e) {
				Assert.fail("Should not happen");
			}
			isAnyJobRunning = yarnClient.getApplications().stream()
				.anyMatch(YarnTestBase::isApplicationRunning);
		}

		if (isAnyJobRunning) {
			final List<String> runningApps = yarnClient.getApplications().stream()
				.filter(YarnTestBase::isApplicationRunning)
				.map(app -> "App " + app.getApplicationId() + " is in state " + app.getYarnApplicationState() + '.')
				.collect(Collectors.toList());
			if (!runningApps.isEmpty()) {
				Assert.fail("There is at least one application on the cluster that is not finished." + runningApps);
			}
		}
	} catch (Throwable t) {
		throw new AssertionError(ExceptionUtils.firstOrSuppressed(t, testFailure));
	}
}

if (testFailure != null) {
	throw new AssertionError(testFailure);
}
```

This is what I meant in my other comment about having to retain both try-catch blocks. We need the outer one to catch the test exception so _can_ suppress later exceptions, and we need the inner one to _do_ the actual suppression.

Reply:
You are right @zentol. What about the following idea: We encapsulate the cleanup logic in an `AutoCloseable` instance.

```
protected void runTest(RunnableWithException test) throws Exception {
		try (final CleanupYarnApplication ignored = new CleanupYarnApplication()){
			test.run();
		}
	}

private class CleanupYarnApplication implements AutoCloseable {
	@Override
	public void close() throws Exception {
		Deadline deadline = Deadline.now().plus(Duration.ofSeconds(10));

		boolean isAnyJobRunning = yarnClient.getApplications().stream()
			.anyMatch(YarnTestBase::isApplicationRunning);

		while (deadline.hasTimeLeft() && isAnyJobRunning) {
			try {
				Thread.sleep(500);
			} catch (InterruptedException e) {
				Assert.fail("Should not happen");
			}
			isAnyJobRunning = yarnClient.getApplications().stream()
				.anyMatch(YarnTestBase::isApplicationRunning);
		}

		if (isAnyJobRunning) {
			final List<String> runningApps = yarnClient.getApplications().stream()
				.filter(YarnTestBase::isApplicationRunning)
				.map(app -> "App " + app.getApplicationId() + " is in state " + app.getYarnApplicationState() + '.')
				.collect(Collectors.toList());
			if (!runningApps.isEmpty()) {
				Assert.fail("There is at least one application on the cluster that is not finished." + runningApps);
			}
		}
	}
}
```